### PR TITLE
Feat/generate population

### PR DESCRIPTION
**Pull request by stmeinert**
added capability to create a initial population for a given instance

### CODE REVIEW

File: src/main/java/de/uos/informatik/ko/rcp/geneticAlgo/GeneratePop.java [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+        int i = 0;
+        for(Integer elem: candidates) {
+            if (i == item) {
+                mycandidate = elem;
+                break;
+            }
+            i++;
+        }
+        returncanidate = mycandidate;
+        return returncanidate;
```
Comment (leon-richardt) [inline src/main/java/de/uos/informatik/ko/rcp/geneticAlgo/GeneratePop.java]: If we cannot switch to a random access collection, I propose this implementation instead: https://stackoverflow.com/a/25410520

Comment (leon-richardt) [reply src/main/java/de/uos/informatik/ko/rcp/geneticAlgo/GeneratePop.java]: We could also consider always choosing the first element that is returned by an iterator into the set. [The documentation](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/HashSet.html#iterator()) says:
> The elements are returned in no particular order.

I'm not yet sure how "good" that randomness is for our use case but it might be good enough to generate an initial population. In particular, this *could* cause all our initial activity lists to be the same. (It would definitely require testing though.)

File: src/main/java/de/uos/informatik/ko/rcp/geneticAlgo/GeneratePop.java [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+            ArrayList<Integer> resultsList = new ArrayList<Integer>();
+            for (Integer elem: scheduled){
+                resultsList.add(elem);
+            }
+            results.add(resultsList);
+            scheduled.clear();
+            candidates.clear();
+            if (results.size() >= pop_size){
+                break;
+            }
```
Comment (leon-richardt) [inline src/main/java/de/uos/informatik/ko/rcp/geneticAlgo/GeneratePop.java:64]: Why is this needed? Shouldn't the loop condition take care of this already?

File: src/main/java/de/uos/informatik/ko/rcp/geneticAlgo/GeneratePop.java [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+                candidates.remove(actcandidate);
+                // delete scheduled activity from predecessor list
+                for (Map.Entry<Integer, HashSet<Integer>> entry : mypredecessors.entrySet()) {
+                    entry.getValue().remove(actcandidate);
+                }
+            }
+            ArrayList<Integer> resultsList = new ArrayList<Integer>();
+            for (Integer elem: scheduled){
+                resultsList.add(elem);
+            }
```
Comment (leon-richardt) [inline src/main/java/de/uos/informatik/ko/rcp/geneticAlgo/GeneratePop.java]: We can use `ArrayList`'s ["copy constructor"](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/ArrayList.html#%3Cinit%3E(java.util.Collection)):
```suggestion
            ArrayList<Integer> resultsList = new ArrayList<Integer>(scheduled);
```

File: src/main/java/de/uos/informatik/ko/rcp/geneticAlgo/GeneratePop.java [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+        }
+        return myArray;
+    }
+    public static int GetRowCount(HashSet<ArrayList<Integer>> mySet){
+        return mySet.size();
+    }
+
+    public static int GetColumnCount(HashSet<ArrayList<Integer>> mySet){
+        return mySet.iterator().next().size();
+    }
```
Comment (leon-richardt) [inline src/main/java/de/uos/informatik/ko/rcp/geneticAlgo/GeneratePop.java:98]: What are these needed for? And why do they start with a capital letter?

File: src/main/java/de/uos/informatik/ko/rcp/geneticAlgo/GeneratePop.java [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+    //change to Hashset of ArrayLists
+    public static HashSet<ArrayList<Integer>> generatePop(Instance instance, Integer pop_size,Random myGenerator){
+        // all activities without predecessor are possible candidates
+        HashSet<ArrayList<Integer>> results = new HashSet<ArrayList<Integer>>();
+        HashMap<Integer,HashSet<Integer>> mypredecessors;
+        HashSet<Integer> candidates = new HashSet<Integer>();
+        //scheduled activities
+        ArrayList<Integer> scheduled = new ArrayList<Integer>();
+        for (int i=0; i< pop_size*5; i++) {
+            mypredecessors = Utils.buildPredecessorMap(instance);
```
Comment (leon-richardt) [inline src/main/java/de/uos/informatik/ko/rcp/geneticAlgo/GeneratePop.java]: Does this need to be rebuilt on every iteration?

File: src/main/java/de/uos/informatik/ko/rcp/geneticAlgo/GeneratePop.java [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+    private GeneratePop() {}
+    //change to Hashset of ArrayLists
+    public static HashSet<ArrayList<Integer>> generatePop(Instance instance, Integer pop_size,Random myGenerator){
+        // all activities without predecessor are possible candidates
+        HashSet<ArrayList<Integer>> results = new HashSet<ArrayList<Integer>>();
+        HashMap<Integer,HashSet<Integer>> mypredecessors;
+        HashSet<Integer> candidates = new HashSet<Integer>();
+        //scheduled activities
+        ArrayList<Integer> scheduled = new ArrayList<Integer>();
+        for (int i=0; i< pop_size*5; i++) {
```
Comment (leon-richardt) [inline src/main/java/de/uos/informatik/ko/rcp/geneticAlgo/GeneratePop.java]: Where does the `5` come from?

**Review by leon-richardt — CHANGES_REQUESTED** [review]

File: src/main/java/de/uos/informatik/ko/rcp/geneticAlgo/GeneratePop.java (hunk 4 above)
Comment (stmeinert) [reply src/main/java/de/uos/informatik/ko/rcp/geneticAlgo/GeneratePop.java:98]: You have a point with the capital letter. :)
I talked to Benedikte and Maria and they wanted to implement there algorithm with an array.
So i created these three functions to give them the chance to use arrays instead of the  HashSet<ArrayList<Integer>>.
Of course, they can use the HashSet, too.

File: src/main/java/de/uos/informatik/ko/rcp/geneticAlgo/GeneratePop.java (hunk 2 above)
Comment (stmeinert) [reply src/main/java/de/uos/informatik/ko/rcp/geneticAlgo/GeneratePop.java:64]: In the current version not. I think there are prettier ways to implement the outer loop. Currently the for loop just consists of pop_size*(a_new_variable_not_five) iterations. If the algorithm generates pop_size unique elements, it should stop iterating, since it already created enough elements.

File: src/main/java/de/uos/informatik/ko/rcp/geneticAlgo/GeneratePop.java (hunk 3 above)
Comment (stmeinert) [reply src/main/java/de/uos/informatik/ko/rcp/geneticAlgo/GeneratePop.java]: You’re right.

File: src/main/java/de/uos/informatik/ko/rcp/geneticAlgo/GeneratePop.java (hunk 5 above)
Comment (stmeinert) [reply src/main/java/de/uos/informatik/ko/rcp/geneticAlgo/GeneratePop.java]: Because of line 46 in the code. I remove elements from the predecessor map. I need a "fresh" version of the predeccessor map in every iteration. But your point is valid, since i think it would be more efficient if we create the complete predecessor map only once and copy it in every new iteration.

File: src/main/java/de/uos/informatik/ko/rcp/geneticAlgo/GeneratePop.java (hunk 6 above)
Comment (stmeinert) [reply src/main/java/de/uos/informatik/ko/rcp/geneticAlgo/GeneratePop.java]: Sorry, magic number.
We should use of course use a variable
The loop defines the number of the attempts to create a new population member.
Since the function could generate the same candidate, we have to investigate a appropriate amount of iterations.

File: src/main/java/de/uos/informatik/ko/rcp/geneticAlgo/GeneratePop.java (hunk 1 above)
Comment (stmeinert) [reply src/main/java/de/uos/informatik/ko/rcp/geneticAlgo/GeneratePop.java]: I like the above mentioned version from stackoverflow.

Concerning your second comment, i think we are loosing the possibility to use the seed, since it isn't used if we just use the iterator right? 
I think it could be helpful if we try to test or debug our code.
And we have, as you mentioned, to test how "deterministic" the iterator is.

Comment (leon-richardt) [reply src/main/java/de/uos/informatik/ko/rcp/geneticAlgo/GeneratePop.java]: Good point about using a seed, didn't consider that :+1:  We should stick to the SO version then IMO

File: src/main/java/de/uos/informatik/ko/rcp/geneticAlgo/GeneratePop.java (hunk 4 above)
Comment (leon-richardt) [reply src/main/java/de/uos/informatik/ko/rcp/geneticAlgo/GeneratePop.java:98]: Aha, makes sense then :thinking: I think that should be something to discuss with the whole group in that case

File: src/main/java/de/uos/informatik/ko/rcp/geneticAlgo/GeneratePop.java (hunk 5 above)
Comment (leon-richardt) [reply src/main/java/de/uos/informatik/ko/rcp/geneticAlgo/GeneratePop.java]: Yeah, I think a copy will be faster for sure :+1:

File: src/main/java/de/uos/informatik/ko/rcp/geneticAlgo/GeneratePop.java (hunk 6 above)
Comment (leon-richardt) [reply src/main/java/de/uos/informatik/ko/rcp/geneticAlgo/GeneratePop.java]: Right, I didn't consider the possibility of creating the same population member twice. I also think that finding a good number of attempts might require some experimentation

File: src/main/java/de/uos/informatik/ko/rcp/geneticAlgo/GeneratePop.java (hunk 2 above)
Comment (leon-richardt) [reply src/main/java/de/uos/informatik/ko/rcp/geneticAlgo/GeneratePop.java:64]: Yep, https://github.com/leon-richardt/rcpsp-solver/pull/5#discussion_r637578921 cleared this up for me :+1: Maybe a comment could be added somewhere explaining that member generation will be attempted again for non-unique activity lists

File: src/main/java/de/uos/informatik/ko/rcp/geneticAlgo/GeneratePop.java (hunk 4 above)
Comment (leon-richardt) [reply src/main/java/de/uos/informatik/ko/rcp/geneticAlgo/GeneratePop.java:98]: (I will keep this unresolved until https://github.com/leon-richardt/rcpsp-solver/issues/8 is closed.)
